server/webhook.go

```diff
@@ -38,145 +45,327 @@ type Webhook struct {
-// posted.
-func (w *Webhook) SlackAttachment() (*model.SlackAttachment, error) {
-	switch w.WebhookEvent {
-	case "jira:issue_created":
-	case "jira:issue_updated":
-		isResolutionChange := false
-		for _, change := range w.ChangeLog.Items {
-			if change.Field == "resolution" {
-				isResolutionChange = (change.FromString == "") != (change.ToString == "")
-				break
-			}
+const (
+	md_RootStyle   = "## "
+	md_UpdateStyle = "###### "
+)
+
+var formatters = map[string]func(w *Webhook) string{
+	"jira:issue_created": mdIssueCreated,
+	"jira:issue_updated": mdIssueUpdated,
+	"comment_created":    mdComment,
+	"comment_updated":    mdComment,
+	"comment_deleted":    mdComment,
+}
+
+func (w *Webhook) Markdown() string {
+	f := formatters[w.WebhookEvent]
+	if f == nil {
+		f = mdDefault
+	}
+	return f(w)
+}
+
+func mdDefault(w *Webhook) string {
+	return ""
+}
+
+func mdIssueCreated(w *Webhook) string {
+	s := md_RootStyle
+	s += fmt.Sprintf("%v created a %v %v", md_User(&w.User), md_IssueType(w), md_IssueLongLink(w))
+	s += "\n"
+
+	s += join(w,
+		md_IssuePriority,
+		md_IssueAssignedTo,
+		md_IssueReportedBy,
+		md_IssueLabels,
+		md_IssueHashtags,
+	)
+	s += "\n"
+	s += "\n"
+
+	s += md_IssueDescription(w)
+
+	return s
+}
+
+func mdIssueUpdated(w *Webhook) string {
+	s := md_UpdateStyle
+	headline := ""
+	extra := ""
+
+	switch w.IssueEventTypeName {
+	case "issue_assigned":
+		assignee := "_nobody_"
+		if w.Issue.Fields.Assignee != nil {
+			assignee = md_User(w.Issue.Fields.Assignee)
 		}
-		if !isResolutionChange {
-			return nil, nil
+		headline = fmt.Sprintf("assigned %v to %v", md_IssueLongLink(w), assignee)
+
+	case "issue_updated",
+		"issue_generic":
+		// edited summary, description, updated priority, status, etc.
+		headline = md_HeadlineFromChangeLog(w)
+		if w.showDescription {
+			extra = md_IssueDescription(w)
 		}
-	case "jira:issue_deleted":
-		if w.Issue.Fields.Resolution != nil {
-			return nil, nil
+	}
+	if headline == "" {
+		// return fmt.Sprintf("```%v```\n", w.RawJSON)
+		return ""
+	}
+
+	s += md_User(&w.User) + " " + headline + " " + md_IssueHashtags(w)
+	s += "\n"
+
+	if extra != "" {
+		s += extra
+	}
+	s += "\n"
+
+	return s
+}
+
+func mdComment(w *Webhook) string {
+	s := md_UpdateStyle
+	headline := ""
+	extra := ""
+
+	switch w.WebhookEvent {
+	case "comment_deleted":
+		headline = fmt.Sprintf("removed a comment from %v", md_IssueLongLink(w))
+
+	case "comment_updated":
+		headline = fmt.Sprintf("edited a comment in %v", md_IssueLongLink(w))
+		extra = w.Comment.Body
+
+	case "comment_created":
+		headline = fmt.Sprintf("commented on %v", md_IssueLongLink(w))
+		extra = w.Comment.Body
+
+	}
+	if headline == "" {
+		return ""
+	}
+
+	s += md_User(&w.Comment.UpdateAuthor) + " " + headline + " " + md_IssueHashtags(w)
+	s += "\n"
+	if extra != "" {
+		s += extra
+	}
+	s += "\n"
+	return s
+}
+
+func md_HeadlineFromChangeLog(w *Webhook) string {
```

Comment:
Some functions are named without the underscore and some have it (eg `mdComment` vs `md_HeadlineFromChangeLog `). Is that an intentional distinction or just accidental?

Reply:
This was intentional (top-level vs "blocks") but thinking about your comment, and being true to the Go style I'll remove the `_`s.